### ISSUE TITLE
Typechecker bug related to include and modules

### ISSUE THREAD

**Issue by craff** [ocaml/ocaml#11414]
Hello,

The attached three files result in what I think is a wrong type error (reducing to 1 or 2 files make the bug disappear).
 Two ways to have the program accepted:
1°) do 'open Type' in test1.ml instead of 'open Pack '(pack.ml only contains one line 'include Type')
2°) rename Test1 at the end of test1.ml in Test2 (or anything else).

[Bug.zip](https://github.com/ocaml/ocaml/files/9070389/Bug.zip)

Thanks again for offering OCaml to the world,
Christophe

**Comment by craff** [comment]
Forgot: I used ocaml 4.14.0

**Comment by Octachron** [comment]
The linked test is:
```ocaml
module Type = struct
  type _ u = ..

  module type B = sig
    type t
    module type R
  end

  module type C = sig
    module F(T:B) : B with type t = T.t
                       and module type R = T.R
  end
end

module Pack = struct
  include Type
end

open Pack

module Test(C:C) =
  struct
    include
      struct
        module Test1 =
          struct
            type t
            module type R  = sig val u : t u end
          end
      end
    module Test1 = C.F(Test1)
  end
```
The issue seems to appear when trying to erase the shadowed `Test1` module in the `Test` functor.
Interestingly, trying to use directly an anonymous module:
```ocaml
module Test'(C:C) =
  struct
    module Test1 = C.F(struct
            type t
            module type R  = sig val u : t u end
          end)
  end
```
leads to a fatal error `nondep_supertype not included in original module type`.

**Comment by gasche** [comment]
The last person I discussed "nondep erasure stuff" with was @trefis; just sayin'.

**Comment by gasche** [comment]
cc @goldfirere: we are considering assigning y'all to this issue

**Comment by goldfirere** [comment]
@stedolan do you have any ideas here? I see `nondep_supertype` and think of you, on our team. (I was unable to download the original zip file through our firewall, but @Octachron's summary is accurate. Just make a `type.ml`, `pack.ml`, and then `test1.ml` file from the contents above and you can repro the problem with three `ocamlc -c`s.)

**Comment by stedolan** [comment]
Apologies, missed this at the time. This bug was fixed by #11879 (which is in 5.1).
